### PR TITLE
Add jax.random.multinomial.

### LINKED ISSUE #13327: Add random.binomial and random.multinomial
Add JAX counterparts of [numpy.random.binomial](https://numpy.org/doc/stable/reference/random/generated/numpy.random.binomial.html) and [numpy.random.multinomial](https://numpy.org/doc/stable/reference/random/generated/numpy.random.multinomial.html) to [jax.random](https://jax.readthedocs.io/en/latest/jax.random.html) package. See https://github.com/google/jax/issues/480#issuecomment-1272587620 for context. A current workaround is using the [JAX substrate](https://www.tensorflow.org/probability/api_docs/python/tfp/substrates/jax) of [TensorFlow Probability](https://www.tensorflow.org/probability):
```python3
from jax import random, numpy as jnp
from tensorflow_probability.substrates import jax as tfp

def binomial(key, n, p, shape=()):
    return tfp.distributions.Binomial(n, probs=p).sample(
        seed=key,
        sample_shape=shape,
    )

def multinomial(key, n, p, shape=()):
    return tfp.distributions.Multinomial(n, probs=p).sample(
        seed=key,
        sample_shape=shape,
    )

key = random.PRNGKey(0)

key, subkey = random.split(key)
print(binomial(subkey, 9, .8, [2, 5]))

key, subkey = random.split(key)
print(multinomial(subkey, 9, jnp.array([.7, .2, .1]), [4]))
```
Output:
```
[[7. 8. 8. 7. 5.]
 [5. 8. 9. 8. 5.]]
[[7. 1. 1.]
 [4. 3. 2.]
 [5. 3. 1.]
 [4. 4. 1.]]
```

### PR DESCRIPTION
**Pull request by carlosgmartin**
Relanding https://github.com/jax-ml/jax/pull/25688 due to https://github.com/jax-ml/jax/pull/26336.

Reduced the value of the `trials` argument to `multinomial` in the tests, and increased the error tolerance accordingly.

Fixes #13327

### CODE REVIEW

**Comment by carlosgmartin** [comment]
@jakevdp Done.

**Comment by AndrewArrasmith** [comment]
Hi, I have used the multinomial sampling code here and run into significant issues with compilation. Since this doesn't seem to have merged into main yet I don't know where the proper place to raise this issue is so I mention it here. (Sorry!) It is also a trivial fix. 

In my use case the multinomial sampling sits at the heart of a large program and gets called _a lot._ The issue is the unrolled `jax.lax.scan` eating huge amounts of memory and massively increasing the compilation time.

If the line that reads
`remainder, counts = lax.scan(f, n, (ratios, keys), unroll=True)`
is simply changed to 
`remainder, counts = lax.scan(f, n, (ratios, keys), unroll=False)`
then the compilation becomes _much_ faster and less memory hungry.

In my test case the overall first run time (jit and the initial evaluation time) goes from 1h 7min 31s to 17.3 s and the memory used goes from 19 GB to 4 GB. 

Is there some reason why unrolling is needed here? I am assuming that the devs here know more about this than I do but it seems pretty bad to me.

**Comment by carlosgmartin** [comment]
@AndrewArrasmith Fixed.

**Comment by AndrewArrasmith** [comment]
Thanks!

**Comment by carlosgmartin** [comment]
@jakevdp Don't merge yet. I realized that this can be done more efficiently with a divide-and-conquer strategy (e.g. [associative scan](https://docs.jax.dev/en/latest/_autosummary/jax.lax.associative_scan.html)), instead of a linear scan. I'm rewriting the implementation accordingly.

**Comment by carlosgmartin** [comment]
Nevermind. My divide-and-conquer strategy doesn't fit into the paradigm of [associative_scan](https://docs.jax.dev/en/latest/_autosummary/jax.lax.associative_scan.html), because each node must tell each child how many trials it receives, before the child tells the parent its frequency table.

OTOH, doing this via simple Python recursion is slow. Related:
- https://github.com/jax-ml/jax/discussions/6719
- https://github.com/jax-ml/jax/discussions/6838

I'd have to create an implementation of an efficient divide-and-conquer harness.

Let's go with the `scan`-based implementation for now.

File: jax/_src/random.py [hunk 1]
```diff
@@ -2661,6 +2661,67 @@ def binomial(
...
+    An array of counts for each outcome with the specified dtype and with shape
+      ``p.shape`` if ``shape`` is None, otherwise ``shape + (p.shape[-1],)``.
+  """
+
+  key, _ = _check_prng_key("multinomial", key)
+  check_arraylike("multinomial", n, p)
+  n, p = promote_dtypes_inexact(n, p)
+
+  if shape is None:
+      shape = p.shape
```
Comment (jakevdp) [inline jax/_src/random.py:2704]: Please use two-space indentation throughout

File: jax/_src/random.py [hunk 2]
```diff
@@ -2661,6 +2661,67 @@ def binomial(
...
+
+  def f(remainder, ratio_key):
+      ratio, key = ratio_key
+      count = binomial(key, remainder, ratio.clip(0, 1))
+      return remainder - count, count
+
+  p = jnp.moveaxis(p, -1, 0)
+
+  remaining_probs = lax.cumsum(p, 0, reverse=True)
+  ratios = p / jnp.where(remaining_probs == 0, 1, remaining_probs)
```
Comment (jakevdp) [inline jax/_src/random.py:2716]: We likely need the double-where trick here for the sake of differentiability (https://docs.jax.dev/en/latest/faq.html#gradients-contain-nan-where-using-where)

Comment (jakevdp) [reply jax/_src/random.py:2716]: Seems like it's OK then – thanks!

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
CI failures look real – PTAL!

**Review by jakevdp — APPROVED** [review]
